hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java

```diff
@@ -0,0 +1,145 @@
+    hoodieHeartbeatClient.stop(instantTime1);
+    await().atMost(5, SECONDS).until(() -> hoodieHeartbeatClient.getHeartbeat(instantTime1).getNumHeartbeats() > 0);
+    Integer numHeartBeats = hoodieHeartbeatClient.getHeartbeat(instantTime1).getNumHeartbeats();
+    assertTrue(numHeartBeats == 1);
+  }
+
+  @Test
+  public void testIsHeartbeatExpired() throws IOException {
+    HoodieHeartbeatClient hoodieHeartbeatClient =
+        new HoodieHeartbeatClient(metaClient.getFs(), metaClient.getBasePath(), heartBeatInterval, numTolerableMisses);
+    hoodieHeartbeatClient.start(instantTime1);
+    hoodieHeartbeatClient.stop(instantTime1);
+    assertFalse(hoodieHeartbeatClient.isHeartbeatExpired(instantTime1));
+  }
+
+  @Test
+  public void testNumHeartbeatsGenerated() {
+    Long heartBeatInterval = 5000L;
+    HoodieHeartbeatClient hoodieHeartbeatClient =
+        new HoodieHeartbeatClient(metaClient.getFs(), metaClient.getBasePath(), heartBeatInterval, numTolerableMisses);
+    hoodieHeartbeatClient.start("100");
+    await().atMost(5, SECONDS).until(() -> hoodieHeartbeatClient.getHeartbeat(instantTime1).getNumHeartbeats() >= 1);
+  }
+
+  @Test
+  public void testDeleteWrongHeartbeat() throws IOException {
+    HoodieHeartbeatClient hoodieHeartbeatClient =
+        new HoodieHeartbeatClient(metaClient.getFs(), metaClient.getBasePath(), heartBeatInterval, numTolerableMisses);
+    hoodieHeartbeatClient.start(instantTime1);
+    hoodieHeartbeatClient.stop(instantTime1);
+    assertFalse(HeartbeatUtils.deleteHeartbeatFile(metaClient.getFs(), basePath, instantTime2));
+  }
+
+  @Disabled
+  public void testConcurrentReadsAndWritesToSameFile() throws InterruptedException {
+    Thread th1 = new Thread(() -> {
+      try {
+        while (true) {
+          OutputStream outputStream = metaClient.getFs().create(new Path(metaClient.getBasePath() + "/test"), true);
+          outputStream.write("testing".getBytes());
+          outputStream.close();
+        }
+      } catch (Exception e) {
+        LOG.info("Caught Exception th1");
+        e.printStackTrace();
+      }
+    });
+
+    Thread th2 = new Thread(() -> {
+      try {
+        while (true) {
+          FileStatus fs = metaClient.getFs().getFileStatus(new Path(metaClient.getBasePath() + "/test"));
+        }
+      } catch (Exception e) {
+        LOG.info("Caught Exception th2");
+        e.printStackTrace();
+      }
+    });
+
+    Thread th3 = new Thread(() -> {
+      try {
+        while (true) {
+          String heartbeat = org.apache.commons.io.IOUtils.toString(metaClient.getFs().open(new Path(metaClient.getBasePath() + "/test")));
+        }
+      } catch (Exception e) {
+        LOG.info("Caught Exception th3");
+        e.printStackTrace();
+      }
+    });
+
+    th1.start();
+    Thread.sleep(100);
+    th2.start();
+    th3.start();
+
+    // To simulate the issues of race conditions when creating & reading file, simply do th1.join; th2.join; th3.join;
```

Comment:
was wondering how these threads actually stop?

Reply:
I left the test disabled since I was just using this for testing. Have enabled the test now and added the join back.